pkg/nsx/services/securitypolicy/builder.go

```diff
@@ -577,20 +577,41 @@ func (service *SecurityPolicyService) buildRuleOutGroup(obj *v1alpha1.SecurityPo
 	return nsxRuleDstGroup, nsxRuleSrcGroupPath, nsxRuleDstGroupPath, nsxProjectShare, nil
 }
 
-func (service *SecurityPolicyService) buildRuleID(obj *v1alpha1.SecurityPolicy, rule *v1alpha1.SecurityPolicyRule, ruleIdx int) string {
+func (service *SecurityPolicyService) buildRuleID(obj *v1alpha1.SecurityPolicy, rule *v1alpha1.SecurityPolicyRule, ruleIdx int, createdFor string) string {
 	serializedBytes, _ := json.Marshal(rule)
-	return fmt.Sprintf("sp_%s_%s_%d", obj.UID, util.Sha1(string(serializedBytes)), ruleIdx)
+	prefix := "sp"
+	if createdFor == common.ResourceTypeNetworkPolicy {
+		prefix = "np"
+	}
+	return fmt.Sprintf("%s_%s_%s_%d", prefix, obj.UID, util.Sha1(string(serializedBytes)), ruleIdx)
 }
 
-func (service *SecurityPolicyService) buildRuleName(obj *v1alpha1.SecurityPolicy, rule *v1alpha1.SecurityPolicyRule, ruleIdx int) string {
+func (service *SecurityPolicyService) buildRuleDisplayName(obj *v1alpha1.SecurityPolicy, rule *v1alpha1.SecurityPolicyRule, ruleIdx int, portIdx int, portAddressIdx int, createdFor string) (string, error) {
+	if createdFor == common.ResourceTypeNetworkPolicy {
+		// The internal security policy rule converted from network policy has its own name generated by the ingress/egress ports.
+		// e.g. input: internal security policy's rule name: TCP.http-UDP.1234-ingress-allow
+		// expand to NSX security policy rules with name TCP.http-UDP.1234.TCP.80-ingress-allow and TCP.http-UDP.1234.UDP.1234-ingress-allow
+		if rule.Ports == nil {
+			return rule.Name, nil
+		}
+		suffixes := []string{common.NetworkPolicyRuleSuffixIngressAllow, common.NetworkPolicyRuleSuffixEgressAllow}
+		ruleName := rule.Name
+		for _, suffix := range suffixes {
+			index := strings.Index(ruleName, suffix)
+			if index != -1 {
+				return fmt.Sprintf("%s.%s-%s", ruleName[:index], service.buildRulePortString(&rule.Ports[portIdx]), suffix), nil
```

Comment:
We already generated truncated name in buildRulePortString, not sure we still need to do it here.

Reply:
done